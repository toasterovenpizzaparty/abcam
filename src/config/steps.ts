import React from "react";
import { StepDataType } from "../redux/reducers/step-reducer/step-reducer";
import { StepPropTypes } from "../types/types";
import { FIELDS } from "./form-types";

const StepOne = React.lazy(() => import("../pages/steps/step-one/step-one"));
const StepTwo = React.lazy(() => import("../pages/steps/step-two/step-two"));
const StepThree = React.lazy(
  () => import("../pages/steps/step-three/step-three")
);
const StepFour = React.lazy(() => import("../pages/steps/step-four/step-four"));

// Create a simple map to connect our step index to a component
export const StepComponentMap: Readonly<
  Record<number, () => React.FC<StepPropTypes>>
> = {
  0: () => StepOne,
  1: () => StepTwo,
  2: () => StepThree,
  3: () => StepFour,
};

export const StepDescriptionMap: Readonly<Record<number, string>> = {
  0: "Title and rating",
  1: "Description",
  2: "Image and description",
  3: "See your review",
};
export const numOfSteps = Object.keys(StepComponentMap).length;

export const getStepComponent = (index: number) =>
  StepComponentMap[index] || StepOne;

export const getStepIndex = (stepIndex: string | undefined) =>
  Math.max(0, (parseInt(stepIndex || "0") || 0) - 1);

export const getField = (state: StepDataType, field: string) =>
  (state?.fields && state?.fields[field]) || "";

export const getNextPageIndex = (index: number) =>
  Math.min(numOfSteps, index + 1);

export const getPreviousPageIndex = (index: number) => Math.max(0, index - 1);

export const getFields = (fields: string[], stepState: StepDataType) => {
  const mappedFields = fields.map((key) => {
    return {
      key,
      value: (stepState?.fields && stepState?.fields[key]) || null,
    };
  });
  if (mappedFields.length) {
    return mappedFields.reduce(
      (acc, field) => ({
        ...acc,
        ...(field.value !== null ? { [field.key]: field.value } : {}),
      }),
      {}
    );
  } else {
    return {};
  }
};

export const getButtonDisabled = (
  stepState: StepDataType,
  componentIndex: number
) => {
  switch (componentIndex) {
    case 0:
      return (
        [FIELDS.TITLE, FIELDS.RATING].filter(
          (field) => !getField(stepState, field)
        ).length > 0
      );
    case 2:
      return (
        [FIELDS.IMAGE].filter((field) => !getField(stepState, field)).length > 0
      );
    default:
      return false;
  }
};
