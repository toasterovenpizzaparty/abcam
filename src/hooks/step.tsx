import { useContext } from "react";
import { StepContext, ActionTypes } from "../providers/step";

export const useStep = (key: string) => {
  const context = useContext(StepContext);

  return (formFieldKey: string) => {
    return {
      getField: () => {
        if (context && context?.state[key]) {
          return (
            context?.state[key] && context?.state[key].formFields[formFieldKey]
          );
        }
        return null;
      },
      saveField: (value: string) => {
        if (context?.dispatch) {
          context.dispatch({
            type: ActionTypes.SAVE_FORM_FIELD,
            formFields: {
              stepKey: key,
              formFieldKey,
              value,
            },
          });
        }
      },
    };
  };
};
