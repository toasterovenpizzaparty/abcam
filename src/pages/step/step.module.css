.wrapper {
  padding: 1.6rem;
  border: 1px solid #ccc;
  border-radius: 0 0 2px 2px;
  margin: 0;
  text-align: center;
  background-color: #fff;
  position: relative;
  box-sizing: border-box;
}

.navigation {
  margin: 1.6rem 0;
  display: flex;
  justify-content: space-between;
}
