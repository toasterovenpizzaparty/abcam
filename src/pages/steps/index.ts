export * from "./step-one/step-one";
export * from "./step-two/step-two";
export * from "./step-three/step-three";
export * from "./step-four/step-four";
