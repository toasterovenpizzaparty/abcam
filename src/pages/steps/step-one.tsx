import React from "react";

export const StepOne: React.FC = () => {
  return <h1>Hello, world!</h1>;
};
