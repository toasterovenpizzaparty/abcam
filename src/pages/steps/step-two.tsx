import React from "react";

export const StepTwo: React.FC = () => {
  return (
    <section data-test-id='step-two'>
      <h1>Hello, world 2!</h1>
    </section>
  );
};
