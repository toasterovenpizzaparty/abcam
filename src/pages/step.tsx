import React from "react";
import { useParams } from "react-router-dom";
import { StepOne } from "./steps";

// Setup our route params type
type StepRouteParamsType = {
  step: string | undefined;
};

// Create a simple map to connect our step index to a component
const StepComponentMap: Readonly<Record<number, React.FC>> = {
  [0]: StepOne,
};

// Return the corresponding component for the current step.
export const StepPage: React.FC = () => {
  const params = useParams<StepRouteParamsType>();
  const index = Math.max(0, parseInt(params?.step || "0") - 1);
  const StepComponent = StepComponentMap[index] || StepOne;
  return <StepComponent />;
};
