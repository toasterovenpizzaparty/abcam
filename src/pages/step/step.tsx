import React, { useCallback, Suspense } from "react";
import { useHistory } from "react-router-dom";
import { useSharedState } from "../../providers/shared-state";
import { useStepSelector } from "../../redux/selectors/step";
import {
  useCreateReviewAction,
  useUpdateReviewAction,
} from "../../redux/actions";
import { useParams } from "react-router-dom";
import { FIELDS } from "../../config/form-types";
import ProgressIndicator from "../../components/ProgressIndicator";
import Title from "../../components/Title";
import Button from "../../components/Button";
import {
  getStepIndex,
  getStepComponent,
  getButtonDisabled,
  numOfSteps,
  getFields,
  getField,
  getNextPageIndex,
  getPreviousPageIndex,
  StepDescriptionMap,
} from "../../config/steps";
import styles from "./step.module.css";

const ErrorMessage = React.lazy(() => import("../../components/ErrorMessage"));

// Setup our route params type
export type StepRouteParamsType = {
  step: string | undefined;
  id?: string | undefined;
};

// Return the corresponding component for the current step.
export const StepPage: React.FC = () => {
  const history = useHistory();
  const params = useParams<StepRouteParamsType>();
  const updateReview = useUpdateReviewAction();
  const createReview = useCreateReviewAction();
  const api = useSharedState().api;
  const stepState = useStepSelector(params?.id || "step-one");
  const index = getStepIndex(params?.step);
  const StepComponent = getStepComponent(index)();

  const isButtonDisabled = api.isLoading || getButtonDisabled(stepState, index);
  const isPreviousButtonVisible = index - 1 > -1;
  const isNextButtonVisible = index + 1 < numOfSteps;

  const navigateToPage = useCallback(
    async (nextPageIndex) => {
      try {
        let id = params?.id;
        if (!index) {
          const response = await createReview(
            getField(stepState, FIELDS.TITLE),
            getField(stepState, FIELDS.RATING)
          );
          if (!response || !response?.data?.id) {
            throw new Error("No ID was returned from the API");
          }
          id = response?.data.id;
        } else {
          // Update a review and navigate
          const isSuccess = await updateReview(
            getFields(Object.values(FIELDS), stepState)
          );
          if (!isSuccess) {
            throw new Error("Could not update the review.");
          }
        }
        history.push(`/step/${nextPageIndex + 1}/id/${id}`);
      } catch (error) {
        console.log(error);
      }
    },
    [stepState, createReview, history, params?.id, updateReview, index]
  );

  const onPressNext = () =>
    getNextPageIndex(index) !== index &&
    navigateToPage(getNextPageIndex(index));

  const onPressPrevious = () =>
    getPreviousPageIndex(index) !== index &&
    navigateToPage(getPreviousPageIndex(index));

  return (
    <>
      <Title>Submit a review</Title>
      <ProgressIndicator
        numSteps={numOfSteps}
        currentStep={index + 1}
        stepDescriptions={StepDescriptionMap}
      />
      <section
        className={styles.wrapper}
        data-test-id={`step-component-${index}`}
      >
        <Suspense fallback={<></>}>
          <ErrorMessage
            data-test-id='step-error-message'
            isVisible={!!api?.error}
            messages={[
              "Something went wrong while trying to save your review.",
            ]}
          />
          <StepComponent
            errorMessage={""}
            isError={false}
            id={params?.id || "step-one"}
          />
        </Suspense>
      </section>
      <nav className={styles.navigation}>
        <Button
          data-test-id='previous-button'
          isDisabled={isButtonDisabled || !isPreviousButtonVisible}
          onClick={onPressPrevious}
        >
          Previous
        </Button>
        <Button
          data-test-id='next-button'
          isDisabled={isButtonDisabled || !isNextButtonVisible}
          onClick={onPressNext}
        >
          Next
        </Button>
      </nav>
    </>
  );
};
export default StepPage;
