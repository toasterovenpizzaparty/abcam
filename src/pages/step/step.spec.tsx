import React from "react";
import { mount } from "enzyme";
import "../../config/enzyme";
import { Provider } from "react-redux";
import store from "../../redux/store";
import { MemoryRouter as Router, Route } from "react-router-dom";
import { StepPage, getStepComponent, getStepIndex } from "./step";
import { StepOne, StepTwo } from "../steps";

describe("<StepPage />", () => {
  it("Renders the correct step", () => {
    const wrapper = mount(
      <Router initialEntries={["/step/2"]}>
        <Provider store={store}>
          <Route path='/step/:step'>
            <StepPage />
          </Route>
        </Provider>
      </Router>
    );
    expect(wrapper.find('[data-test-id="step-component-1"]').exists()).toBe(
      true
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Retrieves the correct component", () => {
    expect(getStepComponent(0)).toBe(StepOne);
    expect(getStepComponent(1)).toBe(StepTwo);
  });

  it("Returns the correct step index", () => {
    expect(getStepIndex(undefined)).toBe(0);
    expect(getStepIndex("Hi")).toBe(0);
    expect(getStepIndex("2")).toBe(1);
  });

  it("Renders a fallback if the step does not exist", () => {
    const wrapper = mount(
      <Router initialEntries={["/step/undefined-123"]}>
        <Provider store={store}>
          <Route path='/step/:step'>
            <StepPage />
          </Route>
        </Provider>
      </Router>
    );

    expect(wrapper.find('[data-test-id="step-component-0"]').exists()).toBe(
      true
    );
    expect(wrapper).toMatchSnapshot();
  });
});
