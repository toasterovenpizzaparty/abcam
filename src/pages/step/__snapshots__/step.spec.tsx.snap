// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StepPage /> Renders a fallback if the step does not exist 1`] = `ReactWrapper {}`;

exports[`<StepPage /> Renders step 3 1`] = `ReactWrapper {}`;

exports[`<StepPage /> Renders step 4 1`] = `ReactWrapper {}`;

exports[`<StepPage /> Renders the correct step 1`] = `ReactWrapper {}`;
